cmd/skopeo/sync.go

```diff
@@ -0,0 +1,594 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/image/copy"
+	"github.com/containers/image/directory"
+	"github.com/containers/image/docker"
+	"github.com/containers/image/docker/reference"
+	"github.com/containers/image/transports"
+	"github.com/containers/image/types"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+type syncOptions struct {
+	global            *globalOptions
+	srcImage          *imageOptions
+	destImage         *imageDestOptions
+	removeSignatures  bool   // Do not copy signatures from the source image
+	signByFingerprint string // Sign the image using a GPG key with the specified fingerprint
+	sourceYaml        bool
+}
+
+type repoDescriptor struct {
+	DirBasePath  string // base path when source is 'dir'
+	TaggedImages []types.ImageReference
+	Context      *types.SystemContext
+}
+
+type tlsVerifyConfig struct {
+	skip bool
+}
+
+type registrySyncCfg struct {
+	Images      map[string][]string
+	Credentials types.DockerAuthConfig
+	TLSVerify   tlsVerifyConfig `yaml:"tls-verify"`
+	CertDir     string          `yaml:"cert-dir"`
+}
+
+type sourceCfg map[string]registrySyncCfg
+
+// Custom unmarshaler for tls-verify, so when user does not specify the TLS
+// verify option, then it is ON by default.
+func (tls *tlsVerifyConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var verify bool
+	if err := unmarshal(&verify); err != nil {
+		return err
+	}
+
+	tls.skip = !verify
+
+	return nil
+}
+
+// Generates a config structure from a YAML file.
+func newSourceConfig(yamlFile string) (cfg sourceCfg, err error) {
+	err = yamlUnmarshal(yamlFile, &cfg)
+	return
+}
+
+// Checks if a given transport is supported by the sync operation.
+func validSyncTransport(transport types.ImageTransport) bool {
+	switch transport {
+	case docker.Transport:
+		return true
+	case directory.Transport:
+		return true
+	}
+
+	return false
+}
+
+// Return a URL object from an input string
+func parseURL(urlString string) (*url.URL, error) {
+	var parsedURL *url.URL
+
+	parsedURL, err := url.Parse(urlString)
+	if err != nil {
+		return parsedURL, errors.WithMessage(err, "Error while parsing source")
+	}
+
+	valid := validSyncTransport(transports.Get(parsedURL.Scheme))
+	if !valid {
+		return parsedURL, errors.New("Invalid transport")
+	}
+
+	return parsedURL, nil
+}
+
+// Given a tranport and an image name (without the transport), returns an
+// ImageReference.
+func getImageReference(transport types.ImageTransport, imgName string) (types.ImageReference, error) {
```

Comment:
If we use `alltransports.ParseImageName(...)` the reference would point to the transport.